### PR TITLE
Expand the coverage of test_addmm and test_addmm_sizes

### PR DESCRIPTION
**Pull request by zasdfgbnm**
- This test is very fast and very important, so it makes no sense in marking it as slowTest
- This test is should also run on CUDA
- This test should check alpha and beta support
- This test should check `out=` support
- manual computation should use list instead of index_put because list is much faster
- precision for TF32 needs to be fixed. Will do it in future PR.

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -16094,43 +16094,55 @@ def _test(use_out, row_major, incx, incy, lda_tail):
...
-            res2 = torch.zeros(10, 25, device=device, dtype=dtype)
-            res2 += M
-            for i in range(10):
-                for j in range(25):
-                    for k in range(50):
-                        res2[i, j] += m1[i, k] * m2[k, j]
-            self.assertEqual(res1, res2, atol=prec, rtol=0)
+        }[dtype]
+
+        if False and dtype.is_complex:  # bug to be fixed in another PR
```
Comment (zasdfgbnm) [inline test/test_torch.py:16109]: To be fixed in https://github.com/pytorch/pytorch/pull/43827

File: test/test_torch.py [hunk 2]
```diff
@@ -16140,17 +16152,26 @@ def test_addmm_sizes(self, device, dtype):
...
-                    res2 += M
+                    res1 = torch.addmm(M, m1, m2, alpha=alpha, beta=beta)
+                    res2 = torch.full_like(res1, math.nan)
+                    torch.addmm(M, m1, m2, alpha=alpha, beta=beta, out=res2)
+                    res3 = (beta * M).cpu().tolist()
                     for i in range(n):
                         for j in range(m):
                             for l in range(k):
-                                res2[i, j] += m1[i, l] * m2[l, j]
+                                res3[i][j] += (alpha * m1[i, l] * m2[l, j]).item()
```
Comment (zasdfgbnm) [inline test/test_torch.py:16171]: `.item` can be removed with https://github.com/pytorch/pytorch/pull/43844

File: test/test_torch.py [hunk 3]
```diff
@@ -16094,43 +16094,55 @@ def _test(use_out, row_major, incx, incy, lda_tail):
...
             _test(use_out, row_major, incx, incy, lda_tail)
 
-    @slowTest
-    @onlyCPU
-    def test_addmm(self, device):
-        dtypes = {
+    @dtypesIfCUDA(*torch.testing.get_all_complex_dtypes(), *torch.testing.get_all_fp_dtypes(include_bfloat16=False))
+    @dtypes(*torch.testing.get_all_complex_dtypes(), *torch.testing.get_all_fp_dtypes())
+    def test_addmm(self, device, dtype):
+        prec = {
```
Comment (ngimel) [inline test/test_torch.py:16100]: can you use @precisionOverrides for this?

File: test/test_torch.py [hunk 4]
```diff
@@ -16094,43 +16094,55 @@ def _test(use_out, row_major, incx, incy, lda_tail):
...
+            beta = 0.8
+
+        M = torch.randn(10, 25).to(device=device, dtype=dtype)
+        m1 = torch.randn(10, 50).to(device=device, dtype=dtype)
+        m2 = torch.randn(50, 25).to(device=device, dtype=dtype)
+        res1 = torch.addmm(M, m1, m2, alpha=alpha, beta=beta)
+        res2 = torch.full_like(res1, math.nan)
+        torch.addmm(M, m1, m2, alpha=alpha, beta=beta, out=res2)
+        res3 = (beta * M).cpu().tolist()
+        for i in range(10):
```
Comment (ngimel) [inline test/test_torch.py:16123]: I'd prefer comparison with numpy here, we are gradually moving testsuite to compare with numpy references. addmm is not equivalent to matmul, but still can easily be done in numpy

File: test/test_torch.py [hunk 5]
```diff
@@ -16094,43 +16094,55 @@ def _test(use_out, row_major, incx, incy, lda_tail):
...
-            m2 = torch.randn(50, 25).to(device=device, dtype=dtype)
-            res1 = torch.addmm(M, m1, m2)
-            res2 = torch.zeros(10, 25, device=device, dtype=dtype)
-            res2 += M
-            for i in range(10):
-                for j in range(25):
-                    for k in range(50):
-                        res2[i, j] += m1[i, k] * m2[k, j]
-            self.assertEqual(res1, res2, atol=prec, rtol=0)
+        M = torch.randn(10, 1).to(device=device, dtype=dtype).expand(10, 25)
```
Comment (ngimel) [inline test/test_torch.py:16132]: these 2 blocks can be refactored to use the same helper?

File: test/test_torch.py (hunk 2 above)
Comment (ngimel) [reply test/test_torch.py:16171]: #43844 is merged. Also, comparison with numpy reference here.

Comment (zasdfgbnm) [reply test/test_torch.py]: fixed

File: test/test_torch.py (hunk 3 above)
Comment (zasdfgbnm) [reply test/test_torch.py]: Fixed. And note that the precision for bfloat16 is bumped to 0.5

File: test/test_torch.py (hunk 4 above)
Comment (zasdfgbnm) [reply test/test_torch.py]: fixed

File: test/test_torch.py (hunk 5 above)
Comment (zasdfgbnm) [reply test/test_torch.py:16228]: fixed

**Review by ngimel — APPROVED** [review]
Thank you! Waiting for CI. 
Not for this PR, but we should also test that with beta=0 nans and infs in M are not propagated, right now I think we are doing it only for empty inputs.

**Comment by ngimel** [comment]
@ailzhang do you know why xla does not inherit precisionOverride here for bfloat16 (0.6) ? Is it ok if we disable bfloat16 on xla?

File: test/test_torch.py [hunk 6]
```diff
@@ -16194,63 +16194,56 @@ def _test(use_out, row_major, incx, incy, lda_tail):
...
+        res3 = (beta * M).to(numpy_dtype).cpu().numpy() + alpha * (
+            m1.to(numpy_dtype).cpu().numpy() @ m2.to(numpy_dtype).cpu().numpy())
+        res3 = torch.from_numpy(res3).to(dtype)
+        self.assertEqual(res1, res2)
+        self.assertEqual(res1, res3)
+
+    @precisionOverride({torch.double: 1e-8, torch.float: 1e-4, torch.bfloat16: 0.6,
+                        torch.half: 1e-1, torch.cfloat: 1e-4, torch.cdouble: 1e-8})
+    @dtypesIfCUDA(*torch.testing.get_all_complex_dtypes(), *torch.testing.get_all_fp_dtypes(include_bfloat16=False))
+    @dtypes(*torch.testing.get_all_complex_dtypes(), *torch.testing.get_all_fp_dtypes())
```
Comment (ngimel) [inline test/test_torch.py:16220]: can you please not include bfloat16 here, and include it only in @dtypesIfCPU so that bfloat16 is not run on XLA?

**Comment by ailzhang** [comment]
cc: @JackCaoG made some changes to bfloat16 test skip, do you have an idea what's the best workaround here?

**Comment by JackCaoG** [comment]
@ailzhang the change was to auto skip all of the float16 tests, we still want to run bfloat16 tests. My guess is that pt/xla has its own precision overwrite for each type and ignore pytorch's precision.

**Comment by ngimel** [comment]
@JackCaoG @ailzhang so what do you guys suggest we do? Currently bfloat16 test is failing with
```
Tensors failed to compare as equal! With rtol=0.001 and atol=0.001, found 182 element(s) (out of 250) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.53125 (8.375 vs. 7.84375), which occurred at index (6, 4).
```
tbh, rtol 0.001 and atol 0.001 seems incredibly low for bfloat16 even under the best of circumstances, typically errors are much larger. And here the test definitely won't pass with such tolerances. Is there a way to override tolerances on the xla side?

**Comment by JackCaoG** [comment]
> @JackCaoG @ailzhang so what do you guys suggest we do? Currently bfloat16 test is failing with
> 
> ```
> Tensors failed to compare as equal! With rtol=0.001 and atol=0.001, found 182 element(s) (out of 250) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.53125 (8.375 vs. 7.84375), which occurred at index (6, 4).
> ```
> 
> tbh, rtol 0.001 and atol 0.001 seems incredibly low for bfloat16 even under the best of circumstances, typically errors are much larger. And here the test definitely won't pass with such tolerances. Is there a way to override tolerances on the xla side?

@ngimel Let me submit a pr on the pt/xla side to disable this test on our end. Ideally we should take pytorch precision overwrite if it is bigger than pt/xla's. I will investigate a bit on this too.

**Comment by zasdfgbnm** [comment]
Looks like GitHub automatically closed this by mistake? I will reopen and rebase.

**Pull request closed without merging**